components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+	if limInt > 0 {
+		d *= -1 // go forwards to hit second limit switch pin
+	}
+
+	err := g.oAx.motor.GoFor(ctx, d*g.oAx.rpm, 0, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	start := time.Now()
+	for {
+		hit, err := g.limitHit(ctx, limInt)
+		if err != nil {
+			return 0, err
+		}
+		if hit {
+			err = g.oAx.motor.Stop(ctx, nil)
+			if err != nil {
+				return 0, err
+			}
+			break
+		}
+
+		elapsed := start.Sub(start)
+		if elapsed > (time.Second * 15) {
+			return 0, errors.New("gantry timed out testing limit")
+		}
+
+		if !utils.SelectContextOrWait(ctx, time.Millisecond*10) {
+			return 0, ctx.Err()
+		}
+	}
+
+	return g.oAx.motor.Position(ctx, nil)
+}
+
+func (g *limitSwitchGantry) homeWithLimSwitch(ctx context.Context, limSwitchPins []string) error {
+	ctx, done := g.oAx.opMgr.New(ctx)
+	defer done()
+	positionA, err := g.testLimit(ctx, 0)
+	if err != nil {
+		return err
+	}
+
+	np := len(g.limitSwitchPins)
+	var positionB float64
+	switch len(limSwitchPins) {
+	case 1:
+		if g.oAx.lengthMm == 0.0 || g.oAx.mmPerRevolution == 0.0 {
+			return errDimensionsNotFound(g.oAx.name, g.oAx.lengthMm, g.oAx.mmPerRevolution)
+		}
+		revPerLength := g.oAx.lengthMm / g.oAx.mmPerRevolution
+		positionB = positionA + revPerLength
+	case 2:
+		positionB, err = g.testLimit(ctx, 1)
+		if err != nil {
+			return err
+		}
+	default:
+		return errBadNumLimitSwitches(g.oAx.name, np)
+	}
+
+	g.oAx.logger.Debugf("positionA: %0.2f positionB: %0.2f", positionA, positionB)
+
+	g.oAx.positionLimits = []float64{positionA, positionB}
+
+	if g.oAx.gantryRange = positionB - positionA; g.oAx.gantryRange == 0 {
+		return errZeroLengthGantry
+	}
+
+	// Go backwards so limit stops are not hit.
+	x := g.oAx.linearToRotational(0.9 * (g.oAx.gantryRange + positionA))
+	err = g.oAx.motor.GoTo(ctx, g.oAx.rpm, x, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (g *limitSwitchGantry) Position(ctx context.Context, extra map[string]interface{}) ([]float64, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.Position(ctx, extra)
+}
+
+func (g *limitSwitchGantry) MoveToPosition(
+	ctx context.Context,
+	positionsMm []float64,
+	worldstate *referenceframe.WorldState,
+	extra map[string]interface{},
+) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.MoveToPosition(ctx, positionsMm, worldstate, extra)
+}
+
+func (g *limitSwitchGantry) Lengths(ctx context.Context, extra map[string]interface{}) ([]float64, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.Lengths(ctx, extra)
+}
+
+func (g *limitSwitchGantry) Stop(ctx context.Context, extra map[string]interface{}) error {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.Stop(ctx, extra)
+}
+
+func (g *limitSwitchGantry) ModelFrame() referenceframe.Model {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.oAx.ModelFrame()
+}
+
+func (g *limitSwitchGantry) CurrentInputs(ctx context.Context) ([]referenceframe.Input, error) {
+	return g.oAx.CurrentInputs(ctx)
```

Comment:
Naive question: should this lock the mutex too? I don't know what `CurrentInputs` does; maybe the answer is obvious if you know that.

Reply:
I agree with this and the previous comment. Either there should be a mutex on the one axis struct that locks on the interior functions and these outer locks should be removed, or this function should make use of the mutex on `g`. The former seems more scalable to me, but I don't know how much of a concern scalability is in the case of gantries so whichever you think is clearer is probably fine

Reply:
I'll remove the outer locks, and add them to the interior functions where appropriate.